File: E2ETestResults/CoreSubset/Bootstrap/Utility.ts
import { BootstrapResponse, SecondaryBootstrapResponse } from './BootstrapResponse';

import 'mscorlib';

import { MetricsContext, MetricsController, MetricsSuites } from './Performance/MetricsController';

import { tsConfig } from 'TypeDefs';

/**
 * Utility functions for Bootstrap code
 */
export class Utility {
  /**
   * Gets the WindowInstance that corresponds to the viz location. If in WG then it's the WG parent frame.
   * Otherwise it's the viz's frame.
   */
  public static get locationWindow(): Object {
    return window.self;
  }

  public static xhrPostJsonChunked(uri: Object, param: string, firstChunkCallback: (bootstrapResponse: BootstrapResponse) => void, secondaryChunkCallback: (secondaryBootstrapResponse: SecondaryBootstrapResponse) => void, errBack: (object: any) => void, asynchronous: boolean): void {
    let xhr: XMLHttpRequest = new XMLHttpRequest();
    xhr.open('POST', uri, asynchronous);
    xhr.setRequestHeader('Accept', 'text/javascript');
    xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
    if (!ss.isNullOrUndefined(tsConfig.sheetId)) {
      xhr.setRequestHeader('X-Tsi-Active-Tab', tsConfig.sheetId);
    }
    let invokeError: (object: any) => void = Utility.getInvokeErrorDelegate(xhr, errBack);
    let byteOffset: number = 0;
    let consumeJSONChunks: () => void = () => {
      let buffer: string = '';
      try {
        buffer = xhr.responseText;
      } catch (e) { }
      let bufferLength: number = buffer.length;
      while (byteOffset < bufferLength) {
        let newData: string = buffer.substr(byteOffset);
        let regex: RegExp = new RegExp('^(\\d+);');
        let match: string[] = newData.match(regex);
        if (!ss.isValue(match)) {
          return;
        }
        let chunkStart: number = match[0].length;
        let chunkLength: number = ss.parseInt(match[1]).ReinterpretAs();
        if (chunkStart + chunkLength > newData.length) {
          return;
        }
        newData = newData.substr(chunkStart, chunkLength);
        let json: any = null;
        try {
          let contextStr: string = 'Parse ' + ((byteOffset === 0) ? 'Primary' : 'Secondary') + ' JSON';
          {
            const mc: MetricsContext = MetricsController.createContext(contextStr, MetricsSuites.Debug);
            try {
              json = ParseJson<any>(newData);
            } finally {
              if (mc) {
                mc.dispose();
              }
            }
          }
        } catch (e) {
          invokeError(new ss.Exception('Invalid JSON'));
        }
        if (byteOffset === 0) {
          firstChunkCallback(json.ReinterpretAs());
        } else {
          secondaryChunkCallback(json.ReinterpretAs());
        }
        byteOffset += chunkStart + chunkLength;
      }
    };
    let intervalID: number = -1;
    let isReceiving: boolean = false;
    let cannotTouchXhrWhileDownloading: boolean = (window.navigator.userAgent.indexOf('MSIE') >= 0 && number.parseFloat(window.navigator.appVersion.split('MSIE ')[1]) < 10);
    xhr.onreadystatechange = () => {
      try {
        if (!cannotTouchXhrWhileDownloading && xhr.readyState === ReadyState.Loading && xhr.status === 200 && !isReceiving) {
          consumeJSONChunks();
          if (intervalID === -1) {
            intervalID = window.setInterval(consumeJSONChunks, 10);
          }
          isReceiving = true;
          return;
        }
        if (xhr.readyState !== ReadyState.Done) {
          return;
        }
        if (intervalID !== -1) {
          window.clearInterval(intervalID);
          intervalID = -1;
        }
        if (Utility.isSuccessStatus(xhr)) {
          consumeJSONChunks();
        } else {
          invokeError(new ss.Exception('Unable to load ' + uri + '; status: ' + xhr.status));
        }
      } catch (ex) {
        if (typeof ss.getType('ss') === 'undefined') {
          xhr.abort();
        } else {
          throw ex;
        }
      }
    };
    try {
      xhr.send(param);
    } catch (e) {
      invokeError(e);
    }
  }

  /**
   * A Simple synchronous GET to an XML-returning URI
   */
  public static xhrGetXmlSynchronous(uri: Object, errBack: (object: any) => void): string {
    let xhr: XMLHttpRequest = new XMLHttpRequest();
    xhr.open('GET', uri, false);
    xhr.setRequestHeader('Accept', 'text/xml');
    try {
      xhr.send();
    } catch (e) {
      Utility.invokeErrorDelegate(xhr, errBack, e);
      return null;
    }
    return xhr.responseText;
  }

  public static xhrPostJson(uri: Object, param: string, callback: (object: any) => void, errBack: (object: any) => void, asynchronous: boolean): void {
    let xhr: XMLHttpRequest = new XMLHttpRequest();
    xhr.open('POST', uri, asynchronous);
    xhr.setRequestHeader('Accept', 'text/javascript');
    xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
    if (ss.isValue(tsConfig.sheetId)) {
      xhr.setRequestHeader('X-Tsi-Active-Tab', tsConfig.sheetId);
    }
    let invokeError: (object: any) => void = Utility.getInvokeErrorDelegate(xhr, errBack);
    xhr.onreadystatechange = () => {
      if (xhr.readyState !== ReadyState.Done) {
        return;
      }
      if (Utility.isSuccessStatus(xhr)) {
        try {
          let json: any = ParseJson<any>(xhr.responseText);
          callback(json);
        } catch (x) {
          invokeError(x);
        }
      } else {
        invokeError(new ss.Exception('Unable to load ' + uri + '; status: ' + xhr.status));
      }
    };
    try {
      xhr.send(param);
    } catch (e) {
      invokeError(e);
    }
  }

  /**
   * Attaches an event handler for 'message' event, and calls the specified handler
   * function when event is received. If the passed handler handles the event (returns true) the
   * event handler will be removed
   * @param eventHandler 
   */
  public static attachOneTimeMessageHandler(eventHandler: (messageEvent: MessageEvent) => boolean): void {
    let messageListener: HtmlEventHandler = null;
    messageListener = (ev: Event) => {
      let e: MessageEvent = <MessageEvent>ev;
      if (eventHandler(e)) {
        if (ss.isValue((<any>window.self).removeEventListener)) {
          window.removeEventListener('message', messageListener, false);
        } else {
          (<any>window.self).detachEvent('onmessage', messageListener);
        }
      }
    };
    if (ss.isValue((<any>window.self).addEventListener)) {
      window.addEventListener('message', messageListener, false);
    } else {
      (<any>window.self).attachEvent('onmessage', messageListener);
    }
  }

  public static doPostMessage(message: any): boolean {
    let success: boolean = false;
    if ('postMessage' in window) {
      try {
        window.parent.postMessage(message, '*');
        success = true;
      } catch { }
    }
    return success;
  }

  public static parseJson<T>(data: string): T {
    if (ss.isNullOrUndefined(data) || typeof data !== 'string') {
      return ss.getDefaultValue(T);
    }
    data = data.trim();
    return JSON.parse(data);
  }

  public static getTopmostWindow(): Object {
    let win: Object = window.self;
    while (ss.isValue(win.parent) && win.parent !== win) {
      win = win.parent;
    }
    return win;
  }

  private static isSuccessStatus(xhr: XMLHttpRequest): boolean {
    let status: number = ss.isValue(xhr.status) ? xhr.status : 0;
    if ((status >= 200 && status < 300) || status === 304 || status === 1223 || (status === 0 && (window.location.protocol === 'file:' || window.location.protocol === 'chrome:'))) {
      return true;
    }
    return false;
  }

  private static invokeErrorDelegate(xhr: XMLHttpRequest, errBack: (object: any) => void, e: ss.Exception): void {
    if (errBack === null) {
      return;
    }
    let invokeError: (object: any) => void = Utility.getInvokeErrorDelegate(xhr, errBack);
    invokeError(e);
  }

  private static getInvokeErrorDelegate(xhr: XMLHttpRequest, errBack: (object: any) => void): (object: any) => void {
    return (err: any) => {
      err = err || new any();
      err.status = xhr.status;
      err.responseText = xhr.responseText;
      errBack(err);
    };
  }
}

File: test/SaltarelleProjectTests/CoreSubset/Bootstrap/Utility.cs
// ---------------------------------------------------------------------------------------------------------------------
// <copyright file="Utility.cs" company="Tableau Software">
//   This file is the copyrighted property of Tableau Software and is protected by registered patents and other
//   applicable U.S. and international laws and regulations.
//
//   Unlicensed use of the contents of this file is prohibited. Please refer to the NOTICES.txt file for further details.
// </copyright>
// ---------------------------------------------------------------------------------------------------------------------

namespace Tableau.JavaScript.Vql.Bootstrap
{
    using System;
    using System.Html;
    using System.Net;
    using System.Net.Messaging;
    using System.Serialization;
    using System.Text.RegularExpressions;
    using TypeDefs;

    /// <summary>
    /// Utility functions for Bootstrap code
    /// </summary>
    // This is public because the test code uses it.
    public static class Utility
    {
        /// <summary>
        /// Gets the WindowInstance that corresponds to the viz location. If in WG then it's the WG parent frame.
        /// Otherwise it's the viz's frame.
        /// </summary>
        public static WindowInstance LocationWindow
        {
            get
            {
                return Window.Self;
            }
        }

        //// ===========================================================================================================
        //// Methods
        //// ===========================================================================================================

        public static void XhrPostJsonChunked(
            URLStr uri,
            string param,
            Action<BootstrapResponse> firstChunkCallback,
            Action<SecondaryBootstrapResponse> secondaryChunkCallback,
            Action<object> errBack,
            bool asynchronous)
        {
            var xhr = new XmlHttpRequest();
            xhr.Open("POST", uri, asynchronous);
            xhr.SetRequestHeader("Accept", "text/javascript");
            xhr.SetRequestHeader("Content-Type", "application/x-www-form-urlencoded");
            if (!Script.IsNullOrUndefined(TsConfig.SheetId))
            {
                xhr.SetRequestHeader("X-Tsi-Active-Tab", TsConfig.SheetId);
            }
            Action<object> invokeError = GetInvokeErrorDelegate(xhr, errBack);

            int byteOffset = 0;
            Action consumeJSONChunks = delegate
            {
                // Check our response text and see if we can parse the size of the next chunk.
                // If so, see if that entire chunk has arrived. If not, do nothing, and wait to be called
                // back again to do the same check. If it has arrive, attempt to parse it as JSON and pass
                // it to the appropriate callback

                string buffer = "";
                try
                {
                    buffer = xhr.ResponseText;
                }
                catch (Exception)
                {
                    // ignore, IE in ready state 3 will throw an exception when accessing ResponseText.
                }

                int bufferLength = buffer.Length;
                while (byteOffset < bufferLength)
                {
                    string newData = buffer.Substr(byteOffset);
                    var regex = new Regex(@"^(\d+);");
                    string[] match = newData.Match(regex);
                    if (!Script.IsValue(match)) { return; }

                    int chunkStart = match[0].Length;
                    int chunkLength = Script.ParseInt(match[1]).ReinterpretAs<int>();
                    if (chunkStart + chunkLength > newData.Length) { return; }

                    newData = newData.Substr(chunkStart, chunkLength);
                    object json = null;
                    try
                    {
                        string contextStr = "Parse " + ((byteOffset == 0) ? "Primary" : "Secondary") + " JSON";
                        using (var mc = MetricsController.CreateContext(contextStr, MetricsSuites.Debug))
                        {
                            json = ParseJson<object>(newData);
                        }
                    }
                    catch (Exception)
                    {
                        // Bad JSON, ignore chunk
                        // $NOTE: The error handler in CommandController relies
                        // on this error text
                        invokeError(new Exception("Invalid JSON"));
                    }

                    if (byteOffset == 0)
                    {
                        firstChunkCallback(json.ReinterpretAs<BootstrapResponse>());
                    }
                    else
                    {
                        secondaryChunkCallback(json.ReinterpretAs<SecondaryBootstrapResponse>());
                    }

                    byteOffset += chunkStart + chunkLength;
                }
            };

            int intervalID = -1;
            bool isReceiving = false;

            bool cannotTouchXhrWhileDownloading = // B64984 Older IEs choke on accessing xhr.Status in ReadyState 3?
                (Window.Navigator.UserAgent.IndexOf("MSIE") >= 0 && // from BrowserSupport::DetectBrowser
                 double.Parse(Window.Navigator.AppVersion.Split("MSIE ")[1]) < 10); // from Dojo

            xhr.OnReadyStateChange = delegate
            {
                try
                {
                    // ReadyState change firing across the browsers is a mess.
                    // IE10 (all IEs?) will fire Receiving every time it gets a new 4KB chunk.
                    // Some other browsers will only fire Receiving once.
                    // Chrome (at least) even fires Loaded more than once.
                    // Every browser fires Receiving, even if you can't read the responseText (IE < 10)
                    // but in that case, we try to be extra safe with cannotTouchXhrWhileDownloading
                    if (!cannotTouchXhrWhileDownloading &&
                        xhr.ReadyState == ReadyState.Loading &&
                        xhr.Status == 200 &&
                        !isReceiving) // we only want to enter this once
                    {
                        consumeJSONChunks();
                        if (intervalID == -1) // extra paranoia to make sure we only set one interval
                        {
                            intervalID = Window.SetInterval(consumeJSONChunks, 10);
                        }
                        isReceiving = true;
                        return;
                    }
                    if (xhr.ReadyState != ReadyState.Done)
                    {
                        return;
                    }

                    if (intervalID != -1)
                    {
                        Window.ClearInterval(intervalID);
                        intervalID = -1;
                    }

                    if (IsSuccessStatus(xhr))
                    {
                        consumeJSONChunks();
                    }
                    else
                    {
                        invokeError(new Exception("Unable to load " + uri + "; status: " + xhr.Status));
                    }
                }
                catch (Exception ex)
                {
                    // BUGZID: 127983 cleanly nav away from embedded viz in IE in single-page app like AngularJS.
                    // IE will continue calling the OnReadyStateChange for any outstanding XHRs, so abort.
                    // detect we're leaving the app/viz based on if javascript global variable has become undefined,
                    // (undefined can truly become 'undefined')
                    if (Script.IsUndefined(Type.GetType("ss")))
                    {
                        xhr.Abort();
                    }
                    else
                    {
                        throw ex;
                    }
                }
            };

            try
            {
                xhr.Send(param);
            }
            catch (Exception e)
            {
                invokeError(e);
            }
        }

        /// <summary>A Simple synchronous GET to an XML-returning URI</summary>
        ///
        internal static string XhrGetXmlSynchronous(
            URLStr uri,
            Action<object> errBack)
        {
            var xhr = new XmlHttpRequest();
            xhr.Open("GET", uri, false);
            xhr.SetRequestHeader("Accept", "text/xml");

            try
            {
                xhr.Send();
            }
            catch (Exception e)
            {
                InvokeErrorDelegate(xhr, errBack, e);
                return null;
            }

            return xhr.ResponseText;
        }

        internal static void XhrPostJson(
            URLStr uri,
            string param,
            Action<object> callback,
            Action<object> errBack,
            bool asynchronous)
        {
            var xhr = new XmlHttpRequest();
            xhr.Open("POST", uri, asynchronous);
            xhr.SetRequestHeader("Accept", "text/javascript");
            xhr.SetRequestHeader("Content-Type", "application/x-www-form-urlencoded");
            if (Script.IsValue(TsConfig.SheetId))
            {
                xhr.SetRequestHeader("X-Tsi-Active-Tab", TsConfig.SheetId);
            }
            Action<object> invokeError = GetInvokeErrorDelegate(xhr, errBack);

            xhr.OnReadyStateChange = delegate
            {
                // we don't care about this
                if (xhr.ReadyState != ReadyState.Done)
                {
                    return;
                }

                if (IsSuccessStatus(xhr))
                {
                    try
                    {
                        object json = ParseJson<object>(xhr.ResponseText);
                        callback(json);
                    }
                    catch (Exception x)
                    {
                        invokeError(x);
                    }
                }
                else
                {
                    invokeError(new Exception("Unable to load " + uri + "; status: " + xhr.Status));
                }
            };

            try
            {
                xhr.Send(param);
            }
            catch (Exception e)
            {
                invokeError(e);
            }
        }

        /// <summary>
        /// Attaches an event handler for 'message' event, and calls the specified handler
        /// function when event is received. If the passed handler handles the event (returns true) the
        /// event handler will be removed
        /// </summary>
        /// <param name="eventHandler"></param>
        internal static void AttachOneTimeMessageHandler(Func<MessageEvent, bool> eventHandler)
        {
            HtmlEventHandler messageListener = null;
            messageListener = (Event ev) =>
            {
                var e = (MessageEvent)ev;
                if (eventHandler(e))
                {
                    if (Script.IsValue(((dynamic)Window.Self).removeEventListener))
                    {
                        Window.RemoveEventListener("message", messageListener, false);
                    }
                    else
                    {
                        ((dynamic)Window.Self).detachEvent("onmessage", messageListener);
                    }
                }
            };

            if (Script.IsValue(((dynamic)Window.Self).addEventListener))
            {
                Window.AddEventListener("message", messageListener, false);
            }
            else
            {
                ((dynamic)Window.Self).attachEvent("onmessage", messageListener);
            }
        }

        public static bool DoPostMessage(object message)
        {
            bool success = false;
            if (Script.In(typeof(Window), "postMessage"))
            {
                try
                {
                    Window.Parent.PostMessage(message, "*");
                    success = true;
                }
                catch
                {
                    // possible to get security exception
                }
            }
            return success;
        }

        internal static T ParseJson<T>(string data)
        {
            // typeof data !== "string" || !data)
            if (Script.IsNullOrUndefined(data) || Script.TypeOf(data) != "string")
            {
                return default(T);
            }

            // Make sure leading/trailing whitespace is removed (IE can't handle it)
            data = data.Trim();

            return Json.Parse<T>(data);
        }

        internal static WindowInstance GetTopmostWindow()
        {
            WindowInstance win = Window.Self;

            while (Script.IsValue(win.Parent) && win.Parent != win)
            {
                win = win.Parent;
            }

            return win;
        }

        private static bool IsSuccessStatus(XmlHttpRequest xhr)
        {
            int status = Script.IsValue(xhr.Status) ? xhr.Status : 0;
            if ((status >= 200 && status < 300) ||
                 status == 304 || // Not Modified

                 // IE misinterpreting HTTP status code 204 as 1223
                 // (http://www.mail-archive.com/jquery-en@googlegroups.com/msg13093.html)
                 status == 1223 ||

                // IE mangled the status code
                (status == 0 && (Window.Location.Protocol == "file:" ||
                                 Window.Location.Protocol == "chrome:")))
            {
                return true;
            }

            return false;
        }

        private static void InvokeErrorDelegate(XmlHttpRequest xhr, Action<object> errBack, Exception e)
        {
            if (errBack == null)
            {
                return;
            }

            Action<object> invokeError = GetInvokeErrorDelegate(xhr, errBack);
            invokeError(e);
        }

        private static Action<object> GetInvokeErrorDelegate(XmlHttpRequest xhr, Action<object> errBack)
        {
            return (dynamic err) =>
            {
                // TFS 927164: Passed in exception could be undefined
                err = err ?? new object();

                err.status = xhr.Status;
                err.responseText = xhr.ResponseText;
                errBack(err);
            };
        }
    }
}
